### DISCUSSION TITLE
Is it possible to generate assets at build time?

### DISCUSSION THREAD

**Discussion by jensmeindertsma** [vercel/next.js#17222] [Help]
In my React code, I have a link to some favicons, and I'd like to re-use the same URL string when linking to the same icons in my `manifest.webmanifest` file.

I've placed the URL's to these favicons in a `constants.ts` file, and I use these in my React code. I'd like to automatically generate a Manifest file from these constants to make changing the favicons location easier. Is this possible?

I've read [Lee Robinson's approach](https://leerob.io/blog/nextjs-sitemap-robots), but this means that either:

- I can't use TypeScript / ES6 / import my constants with Webpack -> _I need access to my `constants.ts` file._
    **OR**
- The `manifest.webmanifest` URL is a serverless function serving the file -> _I'd like to use only static assets._

**What can I do to make this happen?** 

I wonder whether generating this asset through a page like Lee does has any downsides due to Next treating it like a page?

**Comment by leerob** [comment]
> I can't use TypeScript / ES6 / import my constants with Webpack.

Couldn't you write a Node script with TypeScript, which runs at build time? Rough code:

```js
// scripts/generate-manifest.ts
import fs from 'fs';
import { URL } from '../lib/constants';

(async () => {
  // potentially await getting something

  const manifest = `{
    "name": "HackerWeb",
    "short_name": "HackerWeb",
    "start_url": ".",
    "display": "standalone",
    "background_color": "#fff",
    "description": "A simply readable Hacker News app.",
    "icons": [{
        "src": `${URL}/48.png`,
        "sizes": "48x48",
        "type": "image/png"
    }, {
        "src": `${URL}/72.png`,
        "sizes": "72x72",
        "type": "image/png"
    }, {
        "src": `${URL}/96.png`,
        "sizes": "96x96",
        "type": "image/png"
    }, {
        "src": `${URL}/144.png`,
        "sizes": "144x144",
        "type": "image/png"
    }, {
        "src": `${URL}/168.png`,
        "sizes": "168x168",
        "type": "image/png"
    }, {
        "src": `${URL}/192.png`,
        "sizes": "192x192",
        "type": "image/png"
    }],
    }
  `

  fs.writeFileSync('public/manifest.webmanifest', manifest);
})();
```


https://developer.mozilla.org/en-US/docs/Web/Manifest

**Reply by jensmeindertsma** [reply]
Thanks Lee for the quick response! To have it run at build time I think I have to invoke it from `next.config.js`, and I'm unsure whether this will work with TypeScript because of the `module.exports` syntax in `next.config.js` which makes me think it isn't part of the compilation process. I'll try this though, and see whether it works!

**Reply by jensmeindertsma** [reply]
@leerob I can't seem to get `next.config.js` to run the file. The build fails with `Error: Cannot find module './src/scripts/generate-manifest'` when the file has a `.ts` extension, and when turning it into a JS file the build fails with the following:
```
import fs from "fs";
^^^^^^

SyntaxError: Cannot use import statement outside a module
```
My `next.config.js` looks like:
```js
module.exports = {
  webpack: (config, { isServer }) => {
    if (isServer) {
      require('./src/scripts/generate-manifest');
    }

    return config;
  }
};
```
I can give up the `.ts` file extension, but I still want to use the constants that I have stored in a `.ts` file. Perhaps I can try to run the script from `package.json` scripts with `next build`? 

Curious to hear what you think!

**Reply by stefanprobst** [reply]
adding to `scripts` in `package.json` and running with `ts-node` should work. if you have specified `paths` in `tsconfig.json` you'll probably also need `tsconfig-paths` for module resolution.
